### ISSUE TITLE
tech-debt: attention mask computed once and reused across layers assumes static masks

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#96]
## Description

In the Qwen3 observation path, the attention mask is computed **once** from the initial embedding hidden states and reused across all layers. This optimization assumes that:
1. The mask depends only on sequence length (causal mask), not on hidden state values
2. All layers use the same mask
3. Dynamic masking based on hidden states is not needed

The current implementation is correct for decoder-only causal attention models (Qwen3 and LFM2), but if a model uses dynamic masking that depends on hidden state values (e.g., sparse attention patterns, layer-specific masks), the observer would compute incorrect outputs.

## Exact Code Citations

**src/reap/observer.py, lines 100-110**:
```python
default_mask = (
    _attention_mask(
        h,  # embedding output
        sequence_length=tokens.shape[-1],
        mask_fn=None,
    )
    if mask_fn is None
    else None
)

for layer_idx, layer in enumerate(layers):
    mask = (
        default_mask  # <-- reused every layer
        if mask_fn is None
        else _attention_mask(h, sequence_length=tokens.shape[-1], mask_fn=mask_fn)
    )
```

## Impact

Low for currently supported models (Qwen3-MoE, LFM2-MoE), but a latent bug for any future model adapter that:
- Uses adaptive sparse attention
- Has per-layer mask variations
- Requires hidden-state-dependent masking

## Suggested Fix

Document this assumption explicitly in the adapter interface, or recompute the mask per layer (at minimal cost since `mx.eval` would deduplicate).
